libs/libc/stdio/lib_libvsprintf.c

```diff
@@ -911,6 +915,11 @@ static int vsprintf_internal(FAR struct lib_outstream_s *stream,
           size = 1;
           goto str_lpad;
 
+        case 'm': /* Print error message (%m) */
+          pnt = strerror(saved_errno);
+          size = strlen(pnt); /* Adjusting the size is not supported by %m */
+          goto str_lpad;
+
```

Comment:
can you please avoid messing snprintf?

Reply:
i meant that, printf family should not process "%m".

Reply:
The problem is `syslog` uses the internals of the `vnprintf` to do the processing to reduce code duplication. This is widely used across `libc` variants, I attached an example script in the description.

If you run a locally compiled program - no NuttX, just `libc` - that uses `%m` through a `printf` or similar, it will print the errno string value

Reply:
> The problem is `syslog` uses the internals of the `vnprintf` to do the processing to reduce code duplication. This is widely used across `libc` variants, I attached an example script in the description.
> 
> If you run a locally compiled program - no NuttX, just `libc` - that uses `%m` through a `printf` or similar, it will print the errno string value

macOS:
```
spacetanuki% ./a.out
spacetanuki% cc a.c
spacetanuki% ./a.out
printf also supports %m - m%                                                    
spacetanuki% 
```

i'm sure most BSD variants work this way.

Reply:
So curious. In my machine, it works as expected:

```shell
❯ cat <<EOF > print-demo.c
heredoc> #include <stdio.h>              
heredoc> #include <errno.h>              
heredoc> 
heredoc> int main(void) {    
heredoc>        errno = EINVAL;  
heredoc>        printf("Does printf support %%m?: %m\n");
heredoc>        return 0;        
heredoc> }
heredoc> EOF     
❯ gcc -o print-demo print-demo.c
❯ ./print-demo
Does printf support %m?: Invalid argument
```

Reply:
I'm not quite sure what process to follow. What it's clear is using those `printf` statements within the kernel is unsafe, and should be avoided. Literally, the Linux kernel has its own internal implementations of `vsprintf.c` which are considered safe.

`syslog`, `printf`, and friends are intended to be used from user space. If NuttX is mixing things, it's a problem because changes focusing on achieving POSIX compliance may break things on the kernel side, which shouldn't happen.

@yamt, do you have specific examples of NuttX kernel code using those calls? Mainly because a bugfix should be created to actuate on those.

CC/ @patacongo @xiaoxiang781216 @acassis

Reply:
> I'm not quite sure what process to follow. What it's clear is using those `printf` statements within the kernel is unsafe, and should be avoided. Literally, the Linux kernel has its own internal implementations of `vsprintf.c` which are considered safe.
> 
> `syslog`, `printf`, and friends are intended to be used from user space. If NuttX is mixing things, it's a problem because changes focusing on achieving POSIX compliance may break things on the kernel side, which shouldn't happen.

i personally tend to agree. (i guess i'm biased as my background is BSD)

but it's how nuttx has been. that is, use the (mostly) same api for kernel as well.

> @yamt, do you have specific examples of NuttX kernel code using those calls? Mainly because a bugfix should be created to actuate on those.

eg. see include/debug.h. these debug macros are usually backed by syslog. depending on configurations, they can be used from a bit extreme contexts, like interrupt handlers and very early on the boot process.

probably you can avoid errno access depending on the context.
(using `up_interrupt_context` etc)
although i feel it's a bit hacky, it seems like a common practice.

> 
> CC/ @patacongo @xiaoxiang781216 @acassis

Reply:
> So curious. In my machine, it works as expected:

It will work in one case:  (1) The syscall is a simple, linear, sequence of calls,  (3) the caller of the syscall is the main thread of the task group or this is not a KERNEL build).

But this can fail in the KERNEL mode due to MMU mappings that may no be in place to track to access the required TLS data.

The system call like this is the most trivial case.  How about a true kernel thread that has no error number?  How about an interrupt handler that interrupt who knows what.

The logic of this change is dangerous and incorrect and should never have been merged.  Now what are going to do?  There are two choices: fix it or remove it.  Leaving it alone a pretending there is not problem is not acceptable.

Reply:
> The problem is `syslog` uses the internals of the `vnprintf` to do the processing to reduce code duplication. This is widely used across `libc` variants, I attached an example script in the description.

The is not the root cause of the issue. The cause is in the simulated process model. In BSD and Linux, the errno is declared as a simple global variable one per process.  In NuttX, most users use the FLAT model which does not use the MMU and cannot support a true process model.

Instead, the errno is retain in the TLS of the main thread of the process.  Child threads keep a reference in TLS to the main threads TLS.  When you access the errno, you have to chain through the TLS reference to find the errno.

Interrupt handlers and kernel threads, on the other hand do not have TLS and hence to not have an errno.  Attempts to access the non-existent TLS will have unpredictable results.

System calls can have other TLS issues but tl;dr

Reply:
I'd say @patacongo to revert it for the moment. I made the change not expecting to potentially break builds but to add support to a POSIX-compliant modifier. Theoretically, this runs on user space, so breaking kernel space was never an option in my mind.

Reply:
> > probably you can avoid errno access depending on the context.
> > (using `up_interrupt_context` etc)
> 
> That won't tell you if you are running in a kernel thread or if you are in a system call.

sure, besides that, i guess you should check OSINIT_xxx as well.

Reply:
> > > probably you can avoid errno access depending on the context.
> > > (using `up_interrupt_context` etc)
> > 
> > 
> > That won't tell you if you are running in a kernel thread or if you are in a system call.
> 
> sure, besides that, i guess you should check OSINIT_xxx as well.

I think that we need a single new system call that answers the question:  Is this called from the OS.  That could be hard-coded in KERNEL mode but logic in other modes would have to test (1) if you are in an interrupt, (2) if you are sufficiently initialized, (3) if this is a kernel thread, and (4) if we are in a system call.

That, unfortunately could be a pretty weight solution.

Reply:
> > > > probably you can avoid errno access depending on the context.
> > > > (using `up_interrupt_context` etc)
> > > 
> > > 
> > > That won't tell you if you are running in a kernel thread or if you are in a system call.
> > 
> > 
> > sure, besides that, i guess you should check OSINIT_xxx as well.
> 
> I think that we need a single new system call that answers the question: Is this called from the OS. That could be hard-coded in KERNEL mode but logic in other modes would have to test (1) if you are in an interrupt, (2) if you are sufficiently initialized, (3) if this is a kernel thread, and (4) if we are in a system call.
> 
> That, unfortunately, could be a pretty heavyweight solution.

i think it's a good idea, given that we have similar checks here and there.

i'm not sure how the system call check should be when system call is not enabled though.
even in that case, we need some artificial user-kernel boundary if we want to implement a proper signal handling.